File: static/js/traffic-lights.js
/**
 * Модуль логики светофоров для клиентской части.
 * Централизованная логика определения цветов статусов в JavaScript.
 */

// Константы цветов
const TRAFFIC_LIGHT_COLORS = {
    RED: 'red',
    YELLOW: 'yellow',
    GREEN: 'green', 
    GRAY: 'gray'
};

// Статусы файлов
const FILE_STATUSES = {
    ERROR: 'error',
    UNSUPPORTED: 'unsupported',
    CONTAINS_KEYWORDS: 'contains_keywords',
    NO_KEYWORDS: 'no_keywords',
    PROCESSING: 'processing',
    NOT_CHECKED: 'not_checked'
};

/**
 * Определяет цвет светофора для файла.
 * 
 * @param {string} status - статус файла
 * @param {number|null} charCount - количество символов в файле
 * @param {boolean} hasSearchResults - есть ли результаты поиска для файла
 * @param {boolean} searchPerformed - выполнялся ли поиск
 * @returns {string} цвет светофора
 */
function getFileTrafficLightColor(status, charCount = null, hasSearchResults = false, searchPerformed = false) {
    // Красный: не проиндексированные файлы (всегда красный, независимо от поиска)
    if (status === FILE_STATUSES.ERROR || status === FILE_STATUSES.UNSUPPORTED || charCount === 0) {
        return TRAFFIC_LIGHT_COLORS.RED;
    }
    
    // Если поиск был выполнен
    if (searchPerformed) {
        if (hasSearchResults) {
            return TRAFFIC_LIGHT_COLORS.GREEN; // Зелёный: есть совпадения
        } else {
            return TRAFFIC_LIGHT_COLORS.YELLOW; // Жёлтый: нет совпадений, но файл проиндексирован
        }
    }

    // Серый: проиндексированные, но поиск не производился
    if (status === FILE_STATUSES.CONTAINS_KEYWORDS ||
        status === FILE_STATUSES.NO_KEYWORDS ||
        status === FILE_STATUSES.PROCESSING ||
        status === FILE_STATUSES.NOT_CHECKED) {
        return TRAFFIC_LIGHT_COLORS.GRAY;
    }

    return TRAFFIC_LIGHT_COLORS.GRAY; // По умолчанию серый
}

/**
 * Определяет цвет светофора для папки на основе цветов файлов внутри.
 * Логика согласно спецификации инкремента 11:
 * 1. Зеленый: есть хотя бы один файл с зеленым статусом (успешный поиск)
 * 2. Желтый: есть проиндексированные файлы без результатов поиска (желтые), нет зеленых
 * 3. Красный: все файлы не проиндексированы и был поиск
 * 4. Серый: до первого поиска или при сбросе поиска
 * 
 * @param {string[]} fileColors - массив цветов файлов в папке
 * @returns {string} цвет светофора папки
 */
function getFolderTrafficLightColor(fileColors) {
    if (!fileColors || fileColors.length === 0) {
        return TRAFFIC_LIGHT_COLORS.GRAY;
    }

    const hasGreen = fileColors.includes(TRAFFIC_LIGHT_COLORS.GREEN);
    const hasYellow = fileColors.includes(TRAFFIC_LIGHT_COLORS.YELLOW);
    const hasRed = fileColors.includes(TRAFFIC_LIGHT_COLORS.RED);
    const hasGray = fileColors.includes(TRAFFIC_LIGHT_COLORS.GRAY);

    // Логика папки согласно спецификации инкремента 11
    // 1. Зеленый: если есть хотя бы один файл с успешным результатом поиска
    if (hasGreen) {
        return TRAFFIC_LIGHT_COLORS.GREEN;
    }
    
    // 2. Желтый: если есть проиндексированные файлы без результатов поиска, но нет зеленых
    if (hasYellow) {
        return TRAFFIC_LIGHT_COLORS.YELLOW;
    }
    
    // 3. Красный: если все файлы не проиндексированы и был поиск
    if (hasRed && !hasGray && !hasYellow && !hasGreen) {
        return TRAFFIC_LIGHT_COLORS.RED;
    }
    
    // 4. Серый: по умолчанию (до первого поиска или смешанные состояния)
    return TRAFFIC_LIGHT_COLORS.GRAY;
}

/**
 * Проверяет, проиндексирован ли файл.
 * 
 * @param {string} status - статус файла
 * @param {number|null} charCount - количество символов
 * @returns {boolean} true если файл проиндексирован
 */
function isFileIndexed(status, charCount = null) {
    if (status === FILE_STATUSES.ERROR || status === FILE_STATUSES.UNSUPPORTED) {
        return false;
    }
    if (charCount === 0) {
        return false;
    }
    return true;
}

/**
 * Возвращает приоритет для сортировки по цвету светофора.
 * Больше число = выше в списке.
 * 
 * @param {string} color - цвет светофора
 * @returns {number} приоритет сортировки
 */
function getTrafficLightSortPriority(color) {
    const priorityMap = {
        [TRAFFIC_LIGHT_COLORS.GREEN]: 3,   // Файлы с результатами наверх
        [TRAFFIC_LIGHT_COLORS.GRAY]: 2,    // Проиндексированные в середину
        [TRAFFIC_LIGHT_COLORS.YELLOW]: 2,  // Проиндексированные без результатов в середину
        [TRAFFIC_LIGHT_COLORS.RED]: 1      // Неиндексированные вниз
    };
    return priorityMap[color] || 1;
}

/**
 * Проверяет, выполнялся ли поиск.
 * Использует глобальный флаг window.searchWasPerformed, который устанавливается при выполнении поиска.
 * 
 * @returns {boolean} true если поиск выполнялся
 */
function isSearchPerformed() {
    // Проверяем глобальный флаг (устанавливается при выполнении поиска)
    if (window.searchWasPerformed === true) {
        return true;
    }
    
    // Fallback: проверяем наличие видимых результатов поиска
    const visibleResults = document.querySelectorAll('.file-search-results[style*="display: block"]');
    if (visibleResults.length > 0) {
        return true;
    }
    
    // Проверяем наличие атрибутов data-has-results
    const filesWithResults = document.querySelectorAll('[data-has-results="1"]');
    if (filesWithResults.length > 0) {
        return true;
    }
    
    return false;
}

/**
 * Проверяет, есть ли результаты поиска для файла.
 * 
 * @param {string} filePath - путь к файлу
 * @returns {boolean} true если есть результаты
 */
function hasSearchResultsForFile(filePath) {
    // Метод 1: Проверяем атрибут data-has-results (самый надежный)
    const fileWrapper = document.querySelector(`[data-file-path="${CSS.escape(filePath)}"]`);
    if (fileWrapper && fileWrapper.getAttribute('data-has-results') === '1') {
        return true;
    }
    
    // Метод 2: Проверяем наличие видимых результатов поиска с содержимым
    if (fileWrapper) {
        const searchResults = fileWrapper.querySelector('.file-search-results[style*="display: block"]');
        if (searchResults && searchResults.innerHTML.trim() !== '') {
            // Проверяем, что есть реальные сниппеты или блоки результатов
            const hasSnippets = searchResults.querySelector('.per-term-block, .context-snippet, .found-terms');
            return hasSnippets !== null;
        }
    }
    
    return false;
}

// Экспорт для использования в основном коде
if (typeof window !== 'undefined') {
    // Браузерная среда - добавляем в глобальный объект
    window.TrafficLights = {
        getFileTrafficLightColor,
        getFolderTrafficLightColor,
        isFileIndexed,
        getTrafficLightSortPriority,
        isSearchPerformed,
        hasSearchResultsForFile,
        COLORS: TRAFFIC_LIGHT_COLORS,
        STATUSES: FILE_STATUSES
    };
}

// Экспорт для Node.js/тестирования
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        getFileTrafficLightColor,
        getFolderTrafficLightColor,
        isFileIndexed,
        getTrafficLightSortPriority,
        isSearchPerformed,
        hasSearchResultsForFile,
        TRAFFIC_LIGHT_COLORS,
        FILE_STATUSES
    };
}

File: webapp/utils/traffic_lights.py
"""
Модуль для логики светофоров файлов и папок.
Централизованная логика определения цветов статусов.
"""


class TrafficLightLogic:
    """
    Класс для определения цветов светофоров файлов и папок.
    
    Логика светофоров:
    - Красный: не проиндексированные файлы (error, unsupported, char_count=0)
    - Желтый: проиндексированные без совпадений при поиске
    - Зеленый: проиндексированные с совпадениями при поиске  
    - Серый: проиндексированные без поиска
    """
    
    # Константы цветов
    RED = 'red'
    YELLOW = 'yellow' 
    GREEN = 'green'
    GRAY = 'gray'
    
    # Статусы файлов
    STATUS_ERROR = 'error'
    STATUS_UNSUPPORTED = 'unsupported'
    STATUS_CONTAINS_KEYWORDS = 'contains_keywords'
    STATUS_NO_KEYWORDS = 'no_keywords'
    STATUS_PROCESSING = 'processing'
    STATUS_NOT_CHECKED = 'not_checked'
    
    @classmethod
    def get_file_traffic_light_color(cls, status, char_count=None, has_search_results=False, search_performed=False):
        """
        Определяет цвет светофора для файла.
        
        Args:
            status: статус файла (error, unsupported, contains_keywords, etc.)
            char_count: количество символов в файле (None или число)
            has_search_results: есть ли результаты поиска для файла
            search_performed: выполнялся ли поиск
            
        Returns:
            str: цвет светофора ('red', 'yellow', 'green', 'gray')
        """
        # Красный: не проиндексированные файлы
        if status in [cls.STATUS_ERROR, cls.STATUS_UNSUPPORTED] or char_count == 0:
            return cls.RED
        
        # Если поиск был выполнен
        if search_performed:
            if has_search_results:
                return cls.GREEN  # Зелёный: есть совпадения
            else:
                return cls.YELLOW  # Жёлтый: нет совпадений, но файл проиндексирован
        
        # Серый: проиндексированные, но поиск не производился
        if status in [cls.STATUS_CONTAINS_KEYWORDS, cls.STATUS_NO_KEYWORDS, cls.STATUS_PROCESSING]:
            return cls.GRAY
        
        return cls.GRAY  # По умолчанию серый для not_checked
    
    @classmethod
    def get_folder_traffic_light_color(cls, file_colors):
        """
        Определяет цвет светофора для папки на основе цветов файлов внутри.
        
        Логика папки (вторична, определяется после светофоров файлов):
        1. Зелёный - есть хотя бы один зелёный файл (с результатами поиска)
        2. Жёлтый - есть проиндексированные файлы, но ни один не подходит (есть жёлтые, нет зелёных)
        3. Красный - ВСЕ файлы не проиндексированы (все красные)
        4. Серый - все файлы серые (поиск не выполнялся)
        
        Args:
            file_colors: список цветов файлов в папке
            
        Returns:
            str: цвет светофора папки
        """
        if not file_colors:
            return cls.GRAY
        
        has_red = cls.RED in file_colors
        has_green = cls.GREEN in file_colors  
        has_yellow = cls.YELLOW in file_colors
        has_gray = cls.GRAY in file_colors
        
        # Приоритет: зелёный -> жёлтый -> красный -> серый
        # 1. Если есть хотя бы один зелёный файл - папка зелёная
        if has_green:
            return cls.GREEN
        
        # 2. Если есть жёлтые файлы (проиндексированные без совпадений) - папка жёлтая
        if has_yellow:
            return cls.YELLOW
        
        # 3. Если ВСЕ файлы красные (не проиндексированы) - папка красная
        all_red = all(color == cls.RED for color in file_colors)
        if all_red:
            return cls.RED
        
        # 4. Иначе серая (все файлы серые или смесь серых и красных без жёлтых/зелёных)
        return cls.GRAY
    
    @classmethod
    def is_file_indexed(cls, status, char_count=None):
        """
        Проверяет, проиндексирован ли файл.
        
        Args:
            status: статус файла
            char_count: количество символов
            
        Returns:
            bool: True если файл проиндексирован
        """
        if status in [cls.STATUS_ERROR, cls.STATUS_UNSUPPORTED]:
            return False
        if char_count == 0:
            return False
        return True
    
    @classmethod
    def get_sort_priority(cls, color):
        """
        Возвращает приоритет для сортировки по цвету светофора.
        Больше число = выше в списке.
        
        Args:
            color: цвет светофора
            
        Returns:
            int: приоритет сортировки
        """
        priority_map = {
            cls.GREEN: 3,   # Файлы с результатами наверх
            cls.GRAY: 2,    # Проиндексированные в середину
            cls.YELLOW: 2,  # Проиндексированные без результатов в середину
            cls.RED: 1      # Неиндексированные вниз
        }
        return priority_map.get(color, 1)


def get_traffic_light_color_for_js(status, char_count=None, has_search_results=False, search_performed=False):
    """
    Функция-обёртка для использования в JavaScript через вызовы сервера.
    Дублирует логику TrafficLightLogic.get_file_traffic_light_color.
    """
    return TrafficLightLogic.get_file_traffic_light_color(
        status, char_count, has_search_results, search_performed
    )